File: src/plugin/commands/CustomHandler.java
package plugin.commands;

import arc.struct.Seq;
import arc.util.CommandHandler;
import arc.util.CommandHandler.CommandRunner;
import arc.util.Strings;
import mindustry.Vars;
import mindustry.gen.Player;

import plugin.Bundle;
import plugin.utils.Permission;

import static mindustry.Vars.netServer;

public class CustomHandler {
    final CommandHandler handler;
    public final Seq<CommandData> commands = new Seq<CommandData>();

    public CustomHandler(CommandHandler handler) {
        this.handler = handler;
        registerPseudoCommands();
        Vars.netServer.invalidHandler = (player, response) -> {
            var command = response.command;
            if(response.type == CommandHandler.ResponseType.manyArguments){
                //return "[scarlet]Too many arguments. Usage:[lightgray] " + response.command.text + "[gray] " + response.command.paramText;
                return Bundle.get("commands.manyargs", player.locale, command.text, command.paramText);
            }else if(response.type == CommandHandler.ResponseType.fewArguments){
                //return "[scarlet]Too few arguments. Usage:[lightgray] " + response.command.text + "[gray] " + response.command.paramText;
                return Bundle.get("commands.fewargs", player.locale, command.text, command.paramText);
            }else{ //unknown command
                CommandHandler.Command closest = getClosest(response.runCommand);

                if(closest != null){
                    return Bundle.get("commands.didyoumean", player.locale, closest.text);
                }else{
                    return Bundle.get("commands.unknown", player.locale);
                }
            }
        };
    }

    public CommandHandler.Command getClosest(String name, Player player) {
        int minDst = 0;
        CommandHandler.Command closest = null;
        Seq<Permission> perms = Permission.getPerms(player);

        for(CommandHandler.Command command : netServer.clientCommands.getCommandList()){
            if(!perms.contains(getCommand(command.text).permission)) continue;
            int dst = Strings.levenshtein(command.text, name);
            if(dst < 3 && (closest == null || dst < minDst)){
                minDst = dst;
                closest = command;
            }
        }

        return closest;
    }

    public CommandHandler.Command getClosest(String name) {
        int minDst = 0;
        CommandHandler.Command closest = null;

        for(CommandHandler.Command command : netServer.clientCommands.getCommandList()){
            int dst = Strings.levenshtein(command.text, name);
            if(dst < 3 && (closest == null || dst < minDst)){
                minDst = dst;
                closest = command;
            }
        }

        return closest;
    }

    public CommandData getCommand(String name) {
        return commands.find(c->c.name.equals(name));
    }

    public void registerCommand(String name, CommandRunner<Player> runner) {
        registerCommand(name, "", Permission.none, runner);
    }

    public void registerCommand(String name, String args, CommandRunner<Player> runner) {
        registerCommand(name, args, Permission.none, runner);
    }

    public void registerCommand(String name, String args, Permission perm, CommandRunner<Player> runner) {
        // CommandData cd = new CommandData(name, args, perm);
        commands.add(new CommandData(name, args, perm));

        handler.<Player>register(name, args, "", (a, p)->{
            if(!Permission.getPerms(p).contains(perm)) {
                //Bundle.sendMessage("noperms", p);
                var command = getClosest(name, p);
                if(command == null) {
                    Bundle.sendMessage("commands.unknown", p.locale);
                } else {
                    Bundle.sendMessage("commands.didyoumean", p.locale, command.text);
                }
                return;
            }
            runner.accept(a, p);
        });
    }

    public void registerPseudoCommands() {
        addPseudoCommand("t", "<message...>");
        // addPseudoCommand("a", "<message...>", Permission.admin);
        // addPseudoCommand("votekick", "[player] [reason...]");
        // addPseudoCommand("vote", "<y/n/c>");
        // addPseudoCommand("sync", "");
    }

    public void addPseudoCommand(String name, String args) {
        addPseudoCommand(name, args, Permission.none);
    }

    public void addPseudoCommand(String name, String args, Permission perm) {
        commands.add(new CommandData(name, args, perm));
    }

    public static class CommandData {
        public String name, args;
        public Permission permission;

        CommandData(String name, String args, Permission perm) {
            this.name = name;
            this.args = args;
            this.permission = perm;
        }
    }
}

File: src/plugin/commands/CustomHandler.kt
package plugin.commands

import arc.func.Boolf
import arc.struct.Seq
import arc.util.CommandHandler
import arc.util.CommandHandler.CommandResponse
import arc.util.CommandHandler.CommandRunner
import arc.util.Log
import arc.util.Strings
import mindustry.Vars
import mindustry.core.NetServer.InvalidCommandHandler
import mindustry.gen.Player
import plugin.Bundle
import plugin.database.models.Permission

class CustomHandler {
    var handler: CommandHandler? = null
    val commands: Seq<CommandData> = Seq<CommandData>()

    constructor(handler: CommandHandler) {
        this.handler = handler
        registerPseudoCommands()
        Vars.netServer.invalidHandler = InvalidCommandHandler { player: Player, response: CommandResponse ->
            //val command = response.command
            Log.debug("Type " + response.type)
            if (response.type == CommandHandler.ResponseType.manyArguments) {
                //return "[scarlet]Too many arguments. Usage:[lightgray] " + response.command.text + "[gray] " + response.command.paramText;
                return@InvalidCommandHandler Bundle.get(
                    "commands.manyargs",
                    player.locale,
                    response.command.text,
                    response.command.paramText
                )
            } else if (response.type == CommandHandler.ResponseType.fewArguments) {
                //return "[scarlet]Too few arguments. Usage:[lightgray] " + response.command.text + "[gray] " + response.command.paramText;
                return@InvalidCommandHandler Bundle.get(
                    "commands.fewargs",
                    player.locale,
                    response.command.text,
                    response.command.paramText
                )
            } else { //unknown command
                //val closest = getClosest(response.runCommand)
                val closest: CommandHandler.Command? = getClosest(response.runCommand, player)

                if (closest != null) {
                    return@InvalidCommandHandler Bundle.get("commands.didyoumean", player.locale)
                        .replace("{0}", closest.text)
                } else {
                    return@InvalidCommandHandler Bundle.get("commands.unknown", player.locale)
                }
            }
        }
    }

    fun getClosest(name: String, player: Player): CommandHandler.Command? {
        var minDst = 0
        var closest: CommandHandler.Command? = null
        val perms = Permission.getPerms(player)

        for (command in Vars.netServer.clientCommands.commandList) {
            val cmd = getCommand(command.text)
            if (cmd == null || !perms.contains(cmd.permission)) continue
            val dst = Strings.levenshtein(command.text, name)
            if (dst < 3 && (closest == null || dst < minDst)) {
                minDst = dst
                closest = command
            }
        }

        return closest
    }

    fun getClosest(name: String): CommandHandler.Command? {
        var minDst = 0
        var closest: CommandHandler.Command? = null

        for (command in Vars.netServer.clientCommands.commandList) {
            val dst = Strings.levenshtein(command.text, name)
            if (dst < 3 && (closest == null || dst < minDst)) {
                minDst = dst
                closest = command
            }
        }

        return closest
    }

    fun getCommand(name: String?): CommandData? {
        return commands.find { c: CommandData? -> c!!.name == name }
    }

    fun registerCommand(name: String, runner: CommandRunner<Player>) {
        registerCommand(name, "", Permission.None, runner)
    }

    fun registerCommand(name: String, args: String, runner: CommandRunner<Player>) {
        registerCommand(name, args, Permission.None, runner)
    }

    fun registerCommand(name: String, args: String, perm: Permission, runner: CommandRunner<Player>) {
        // CommandData cd = new CommandData(name, args, perm);
        commands.add(CommandData(name, args, perm))

        handler!!.register(name, args, "", CommandRunner { a: Array<String?>?, p: Player ->
            if (!Permission.getPerms(p).contains(perm)) {
                //Bundle.sendMessage("noperms", p);
                val command = getClosest(name, p)
                if (command == null) {
                    Bundle.sendMessage("commands.unknown", p)
                } else {
                    Bundle.sendMessage("commands.didyoumean", p, command.text)
                }
                return@CommandRunner
            }
            runner.accept(a, p)
        })
    }

    fun registerPseudoCommands() {
        addPseudoCommand("t", "<message...>")
        // addPseudoCommand("a", "<message...>", Permission.admin);
        // addPseudoCommand("votekick", "[player] [reason...]");
        // addPseudoCommand("vote", "<y/n/c>");
        // addPseudoCommand("sync", "");
        handler?.commandList?.each { command ->
            addPseudoCommand(command.text, command.paramText ?: "")
        }
    }

    fun addPseudoCommand(name: String, args: String) {
        addPseudoCommand(name, args, Permission.None)
    }

    fun addPseudoCommand(name: String, args: String, perm: Permission?) {
        commands.add(CommandData(name, args, perm))
    }

    class CommandData internal constructor(var name: String, var args: String, var permission: Permission?) {
        fun getDesc(p: Player): String {
            val req = "commands." + this.name + ".description"
            var desc = Bundle.get(req, p.locale)
            if (desc == req) desc = Bundle.get("commands.nodesc", p.locale)
            return desc
        }
    }
}
